src/lib.rs

```diff
@@ -1237,6 +1392,53 @@ pub async fn build(docker: &Docker, config: &Config) -> Result<(), Error> {
         }
     }
 
+    // Generate {appservice}.yml for appservices running on the host.
+    let generate_appservice_yml =
+        |appservice: &AppServiceConfig, host: &str, path: &Path| -> Result<(), Error> {
+            let mut config = appservice.clone();
+            config.url.set_host(Some(host))?;
+            let writer = std::fs::File::create(&path)
+                .with_context(|| format!("Failed to create file {:?}", path))?;
+            serde_yaml::to_writer(writer, &config)
+                .with_context(|| format!("Failed to write file {:?}", path))?;
+            Ok(())
+        };
+
+    let guest_appservices_root = synapse_root.join("appservices");
+    let host_appservices_root = config.generated_user_files_dir().join("appservices");
+
+    // The (guest) Synapse sees them as running on `host.docker.internal`,
+    // while the (host) AppService see themselves as running on `localhost`.
+    for appservice in &config.appservices.host {
+        // Guest file
+        let guest_path = guest_appservices_root
+            .join(appservice.name.as_ref())
+            .with_extension("yml");
+        generate_appservice_yml(appservice, "host.docker.internal", &guest_path)?;
+        // Host file
+        let host_path = host_appservices_root
+            .join(appservice.name.as_ref())
+            .with_extension("yml");
+        generate_appservice_yml(appservice, "localhost", &host_path)?;
```

Comment:
So we generate twice in two different directories?

Reply:
Yeah, since the contents need to be different (see above).

Reply:
oki